### PR TITLE
Add a backend method to add notes in bulk

### PR DESCRIPTION
**Pull request by abdnh**
Using col.add_note() in a CollectionOp where custom undo entries are involved can easily trigger the "target undo op not found" error due to the undo queue limit. Currently, one has to resort to workarounds such as in the example in https://github.com/ankitects/anki/issues/2628#issuecomment-1703827442

This PR adds the col.add_notes() method to add a list of notes in bulk, registering a single undo entry.

### CODE REVIEW

**Comment by abdnh** [comment]
Hmm, isn't Pylint's `invalid-name` warning disabled? Any idea why it's being reported?

**Comment by dae** [comment]
Not sure what's going on there, but renaming it to idx is probably the fastest solution :-)

**Comment by dae** [comment]
How about using `repeated AddNoteRequest note_and_deck_id = 1` instead of passing a list of notes but a single deck id? That is more flexible, and should require fewer code changes.

File: rslib/src/notes/mod.rs [hunk 1]
```diff
@@ -333,21 +337,52 @@ impl Collection {
...
     ) -> Result<()> {
-        self.canonify_note_tags(note, ctx.usn)?;
-        note.prepare_for_update(ctx.notetype, normalize_text)?;
-        note.set_modified(ctx.usn);
-        self.add_note_only_undoable(note)?;
-        self.generate_cards_for_new_note(ctx, note, did)?;
-        self.set_last_deck_for_notetype(note.notetype_id, did)?;
-        self.set_last_notetype_for_deck(did, note.notetype_id)?;
-        self.set_current_notetype_id(note.notetype_id)
+        let mut notetypes: HashMap<NotetypeId, Arc<Notetype>> = HashMap::new();
```
Comment (dae) [inline rslib/src/notes/mod.rs:347]: This is unnecessary; notetypes are already cached by get_notetype()

File: rslib/src/notes/mod.rs [hunk 2]
```diff
@@ -333,21 +337,52 @@ impl Collection {
...
+        for note in notes.iter() {
+            if !notetypes.contains_key(&note.notetype_id) {
+                let nt = self
+                    .get_notetype(note.notetype_id)?
+                    .or_invalid("missing note type")?;
+                notetypes.insert(note.notetype_id, nt);
+            }
+        }
+        let mut contexts: HashMap<NotetypeId, CardGenContext<&Notetype>> = HashMap::new();
```
Comment (dae) [inline rslib/src/notes/mod.rs:356]: These are fairly cheap to create; I wouldn't bother caching them.

File: rslib/src/notes/mod.rs [hunk 3]
```diff
@@ -333,21 +337,52 @@ impl Collection {
...
+            let ctx = contexts.get(&note.notetype_id).unwrap();
+            self.generate_cards_for_new_note(ctx, note, did)?;
+        }
+        for notetype_id in notetypes.into_keys() {
+            self.set_last_deck_for_notetype(notetype_id, did)?;
+            self.set_last_notetype_for_deck(did, notetype_id)?;
+        }
+        if let Some(last_note) = notes.last() {
+            self.set_current_notetype_id(last_note.notetype_id)?;
+        }
```
Comment (dae) [inline rslib/src/notes/mod.rs:383]: These are quite cheap too. It's great that you're aiming to be efficient here, but in this particular case I think there probably wouldn't be a big difference between this code and just doing the equivalent of `for note_and_deck_id in note_and_deck_ids { old_code() }`. But happy to be proved wrong with a benchmark :-)

File: rslib/src/notes/mod.rs [hunk 4]
```diff
@@ -333,21 +337,52 @@ impl Collection {
...
-        self.canonify_note_tags(note, ctx.usn)?;
-        note.prepare_for_update(ctx.notetype, normalize_text)?;
-        note.set_modified(ctx.usn);
-        self.add_note_only_undoable(note)?;
-        self.generate_cards_for_new_note(ctx, note, did)?;
-        self.set_last_deck_for_notetype(note.notetype_id, did)?;
-        self.set_last_notetype_for_deck(did, note.notetype_id)?;
-        self.set_current_notetype_id(note.notetype_id)
+        let mut notetypes: HashMap<NotetypeId, Arc<Notetype>> = HashMap::new();
+        for note in notes.iter() {
```
Comment (dae) [inline rslib/src/notes/mod.rs:348]: You can do `in &notes` and `in &mut notes` instead of calling iter(mut)

File: rslib/src/notes/mod.rs (hunk 3 above)
Comment (abdnh) [reply rslib/src/notes/mod.rs:383]: Yes. I did a quick test and there's no big difference (~0.2s when adding 10K notes).

**Comment by abdnh** [comment]
> How about using `repeated AddNoteRequest note_and_deck_id = 1` instead of passing a list of notes but a single deck id? That is more flexible, and should require fewer code changes.

Should this be exposed as `def add_notes(self, notes: Sequence[Note], dids: Sequence[DeckId]) -> OpChanges:`? I'd find this slightly inconvenient when I want to add all notes to the same deck (the most common case, I think). What about exposing a different `add_notes_to_decks` function that has this signature?

**Comment by dae** [comment]
What about a public export of AddNoteRequest = AddNoteRequest at the top of collection.py, and changing the signature to add_notes(self, requests: Iterable[AddNoteRequest]). If you have a list of notes, it's just a case of `(AddNoteRequest(note=n, deck_id) for n in notes)`. If you wanted to make things even easier, you could accept [Note | AddNoteRequest] and allow passing in a deck_id as an optional second arg to be used whenever a bare note is provided.

**Comment by abdnh** [comment]
> What about a public export of AddNoteRequest = AddNoteRequest at the top of collection.py, and changing the signature to add_notes(self, requests: Iterable[AddNoteRequest]). If you have a list of notes, it's just a case of (AddNoteRequest(note=n, deck_id) for n in notes). 

But the caller would need to call note._to_backend_note() in this case, right? Should it be marked as public?

**Comment by dae** [comment]
Good point - in that case I'd recommend a dataclass that takes a native note, which we then convert in the function.

File: proto/anki/notes.proto [hunk 5]
```diff
@@ -62,6 +63,15 @@ message AddNoteResponse {
   int64 note_id = 2;
 }
 
+message AddNotesRequest {
+  repeated AddNoteRequest note_and_deck_id = 1;
```
Comment (dae) [inline proto/anki/notes.proto:67]:
```suggestion
  repeated AddNoteRequest note_and_deck_ids = 1;
```

Protobuf convention

Comment (dae) [reply proto/anki/notes.proto:67]: or we could just call it requests

**Review by dae — COMMENTED** [review]
Thanks Abdo - just one minor naming tweak, and it should be good to go.

**Pull request merged**